# <center>Designing & Creating a Normalized Database</center>
## <center>Major League Baseball Data from Retrosheet</center>

## Part I - Getting to Know the Data (EDA)

In [1]:
#1 - read files
import pandas as pd

game=pd.read_csv('game_log.csv')
park=pd.read_csv('park_codes.csv')
person=pd.read_csv('person_codes.csv')
team=pd.read_csv('team_codes.csv')

pd.set_option('max_columns',180)
pd.set_option('max_rows',200000)
pd.set_option('max_colwidth',5000)

/dataquest/system/env/python3/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (12,13,14,15,19,20,81,82,83,84,85,86,87,88,93,94,95,96,97,98,99,100,105,106,108,109,111,112,114,115,117,118,120,121,123,124,126,127,129,130,132,133,135,136,138,139,141,142,144,145,147,148,150,151,153,154,156,157,160) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
#1 - text file explaining fields from main file - game
!cat game_log_fields.txt

Field(s)  Meaning
    1     Date in the form "yyyymmdd"
    2     Number of game:
             "0" -- a single game
             "1" -- the first game of a double (or triple) header
                    including seperate admission doubleheaders
             "2" -- the second game of a double (or triple) header
                    including seperate admission doubleheaders
             "3" -- the third game of a triple-header
             "A" -- the first game of a double-header involving 3 teams
             "B" -- the second game of a double-header involving 3 teams
    3     Day of week  ("Sun","Mon","Tue","Wed","Thu","Fri","Sat")
  4-5     Visiting team and league
    6     Visiting team game number
          For this and the home team game number, ties are counted as
          games and suspended games are counted from the starting
          rather than the ending date.
  7-8     Home team and league
    9     Home team game number
10-11     Visiting and home team score (unquoted)


In [3]:
#1 - game_log explore data
print('game (r,c) ' + str(game.shape))
game.tail(10)

game (r,c) (171907, 161)


date  number_of_game day_of_week v_name v_league  v_game_number  \
171897  20161002               0         Sun    OAK       AL            162   
171898  20161002               0         Sun    TBA       AL            162   
171899  20161002               0         Sun    SDN       NL            162   
171900  20161002               0         Sun    DET       AL            161   
171901  20161002               0         Sun    CHN       NL            162   
171902  20161002               0         Sun    MIL       NL            162   
171903  20161002               0         Sun    NYN       NL            162   
171904  20161002               0         Sun    LAN       NL            162   
171905  20161002               0         Sun    PIT       NL            162   
171906  20161002               0         Sun    MIA       NL            161   

       h_name h_league  h_game_number  v_score  h_score  length_outs  \
171897    SEA       AL            162        3        2         54.0   
171898    TEX       AL            162        6        4         60.0   
171899    ARI       NL            162        2        3         53.0   
171900    ATL       NL            161        0        1         51.0   
171901    CIN       NL            162        7        4         54.0   
171902    COL       NL            162        6        4         60.0   
171903    PHI       NL            162        2        5         51.0   
171904    SFN       NL            162        1        7         51.0   
171905    SLN       NL            162        4       10         51.0   
171906    WAS       NL            162        7       10         51.0   

       day_night completion forefeit protest park_id  attendance  \
171897         D        NaN      NaN     NaN   SEA03     24856.0   
171898         D        NaN      NaN     NaN   ARL02     37015.0   
171899         D        NaN      NaN     NaN   PHO01     31385.0   
171900         D        NaN      NaN     NaN   ATL02     51220.0   
171901         D        NaN      NaN     NaN   CIN09     32587.0   
171902         D        NaN      NaN     NaN   DEN02     27762.0   
171903         D        NaN      NaN     NaN   PHI13     36935.0   
171904         D        NaN      NaN     NaN   SFO03     41445.0   
171905         D        NaN      NaN     NaN   STL10     44615.0   
171906         D        NaN      NaN     NaN   WAS11     28730.0   

        length_minutes v_line_score h_line_score  v_at_bats  v_hits  \
171897           154.0    102000000    000020000       31.0     7.0   
171898           240.0   0210001002   0101000110       42.0    13.0   
171899           157.0    000100100    010000011       33.0     7.0   
171900           167.0    000000000    10000000x       31.0     5.0   
171901           198.0    001100014    300100000       35.0     9.0   
171902           203.0   0200000202   1100100010       39.0    10.0   
171903           159.0    000001100    00100031x       33.0     8.0   
171904           184.0    000100000    23000002x       30.0     4.0   
171905           192.0    000020200    00100360x       35.0     9.0   
171906           216.0    000230020    03023002x       38.0    14.0   

        v_doubles  v_triples  v_homeruns  v_rbi  v_sacrifice_hits  \
171897        2.0        0.0         1.0    3.0               0.0   
171898        6.0        0.0         0.0    5.0               0.0   
171899        4.0        0.0         0.0    2.0               1.0   
171900        1.0        0.0         0.0    0.0               0.0   
171901        2.0        0.0         2.0    7.0               0.0   
171902        4.0        1.0         2.0    6.0               0.0   
171903        3.0        0.0         0.0    2.0               0.0   
171904        0.0        0.0         0.0    1.0               0.0   
171905        0.0        0.0         1.0    4.0               0.0   
171906        1.0        1.0         2.0    7.0               1.0   

        v_sacrifice_flies  v_hit_by_pitch  v_walks  v_inte

- This is the main data set and includes all the details of the game such as:
    - The date, location, time, duration
    - The stats about the team and players

<center>**Columns 106 - 132 visiting team defensive positions <br>
Columns 133 - 159 home team defensive positions <br>
Positions listed [(1-9)](https://en.wikipedia.org/wiki/Baseball_positions)**</center>


| Number | Position |
| --- | --- |
| 1 | pitcher |
| 2 | catcher |
| 3 | first base |
| 4 | second base |
| 5 | third base |
| 6 | shortstop |
| 7 | left field |
| 8 | center field |
| 9 | right field |

In [4]:
#same leagues for h_league
game['v_league'].value_counts()

NL    88866
AL    74713
AA     5039
FL     1243
PL      532
UA      428
Name: v_league, dtype: int64

<center>**Columns 5 visiting team league <br>
Columns 8 home team league <br>
[Full Names](https://en.wikipedia.org/wiki/Baseball_positions)**</center>

| Abbreviation | Full Name | League Type | Status | 
| --- | --- | --- | --- |
| NL | National League | Major | Active |
| AL | American League |Major | Active |
| AA | American Association |Minor | Active |
| FL | Florida Leagueino | Minor | Active |
| PL | Players' League | Major | Terminated |
| UA | Union Association | Major | Terminated |

In [5]:
#1 - park_codes explore data
print('park (r,c) ' + str(park.shape))
park.head()

park (r,c) (252, 9)


park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN

- This data is about the name, location, league and duration of the park

In [6]:
#1 - person_codes explore data
print('person (r,c) ' + str(person.shape))
person.head()

person (r,c) (20494, 7)


id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004       NaN         NaN       NaN
1  aaroh101    Aaron    Hank   04/13/1954       NaN         NaN       NaN
2  aarot101    Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3  aased001     Aase     Don   07/26/1977       NaN         NaN       NaN
4  abada001     Abad    Andy   09/10/2001       NaN         NaN       NaN

- This data is about the names, ID's and start date of people in the league such as players, manager, coaches and umpires

In [7]:
#1 - team_codes explore data
print('team (r,c) ' + str(team.shape))
ny_teams=team[team['city']=='New York']
ny_teams.head(10)

team (r,c) (150, 8)


team_id league  start   end      city       nickname franch_id  seq
9       NYA     AL   1903     0  New York        Yankees       BLA    2
85      NY2    NaN   1871  1875  New York        Mutuals       NY2    1
86      NY3     NL   1876  1876  New York        Mutuals       NY2    2
87      NY4     AA   1883  1887  New York  Metropolitans       NY4    1
88      NYN     NL   1962     0  New York           Mets       NYN    1
89      NYP     PL   1890  1890  New York         Giants       NYP    1
131     NY1     NL   1883  1957  New York         Giants       TRN    2

- This data is about the teams in the league
    - If there is a 0 in the end column it means the team still exists

## Part II - Importing Data into SQLite

### Creating Command and Query Functions

In [17]:
import sqlite3

In [13]:
def run_command(c):
    with sqlite3.connect('mlb.db') as conn:
        conn.isolation_level = None
        conn.execute(c)

In [14]:
def run_query(q):
    with sqlite3.connect('mlb.db') as conn:
        return pd.read_sql(q,conn)

### Creating Tables in SQLite DataBase

In [22]:
#only run once
with sqlite3.connect('mlb.db') as conn:
    game.to_sql('game_log',conn,index=False)
    park.to_sql('park_codes',conn,index=False)
    person.to_sql('person_codes',conn,index=False)
    team.to_sql('team_codes',conn,index=False)

In [23]:
run_query('''SELECT * FROM game_log LIMIT 3''')

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1     None              1    FW1   
1  18710505               0         Fri    BS1     None              1    WS3   
2  18710506               0         Sat    CL1     None              2    RC1   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0     None              1        0        2         54.0         D       None   
1     None              1       20       18         54.0         D       None   
2     None              1       12        4         54.0         D       None   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0     None    None   FOR01       200.0           120.0    000000000   
1     None    None   WAS01      5000.0           145.0    107000435   
2     None    None   RCK01      1000.0           140.0    610020003   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   

  v_intentional_walks  v_strikeouts  v_stolen_bases v_caught_stealing  \
0                None           6.0             1.0              None   
1                None           5.0             3.0              None   
2                None           1.0             0.0              None   

   v_grounded_into_double v_first_catcher_interference  v_left_on_base  \
0                    -1.0                         None             4.0   
1                    -1.0                         None            12.0   
2                    -1.0                         None            10.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   

   v_double_plays  v_triple_plays  h_at_bats  h_hits  h_doubles  h_triples  \
0             0.0             0.0       31.0     4.0        1.0        0.0   
1             2.0             0.0       49.0    14.0        2.0        0.0   
2             0.0             0.0       36.0     7.0        2.0        1.0   

   h_homeruns  h_rbi  h_sacrifice_hits  h_sacrifice_flies  h_hit_by_pitch  \
0         0.0    2.0               0.0                0.0             0.0   
1         0.0   11.0               0.0                0.0             0.0   
2         0.0    2.0               0.0                0.0             0.0   

   h_walks h_intentional_walks  h_strikeouts  h_stolen_bases  \
0      1.0                None           0.0             0.0   
1     10.0                None           2.0             1.0   
2      0.0                None           3.0             5.0   

  h_caught_stealing  h_grounded_into_double h_first_catcher_interference  \
0              None                    -1.0                         None   
1              None                    -1.0                         None   
2              None                    -1.0                         None   

   h_left_on_base  h_pitchers_used  h_individual_earned_runs  \
0             3.0              1.0                       0.0   
1 

### Craeting New Columns in Tables

In [ ]:
c_newcol=
ALTER TABLE game_log ADD COLUMN game_id
UPDATE game_log SET game_id = date+h_name+number_of_game